### PR TITLE
ability for addCommand to add to element scope

### PR DESCRIPTION
**Pull request by abjerstedt**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [X] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/webdriverio/src/commands/browser/$$.js [hunk 1]
```diff
@@ -42,14 +42,15 @@
...
+import merge from 'lodash.merge'
 
 import { findElements, getPrototype as getWDIOPrototype, getElementFromResponse } from '../../utils'
 import { elementErrorHandler } from '../../middlewares'
 import { ELEMENT_KEY } from '../../constants'
 
 export default async function $$ (selector) {
     const res = await findElements.call(this, selector)
-    const prototype = Object.assign({}, this.__propertiesObject__, getWDIOPrototype('element'), { scope: 'element' })
+    const prototype = merge({}, this.__propertiesObject__, getWDIOPrototype('element'), { scope: 'element' })
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/$$.js]: with that we don't need the first param anymore

File: packages/webdriverio/src/commands/browser/$.js [hunk 2]
```diff
@@ -42,14 +42,15 @@
...
+import merge from 'lodash.merge'
 
 import { findElement, getPrototype as getWDIOPrototype, getElementFromResponse } from '../../utils'
 import { elementErrorHandler } from '../../middlewares'
 import { ELEMENT_KEY } from '../../constants'
 
 export default async function $ (selector) {
     const res = await findElement.call(this, selector)
-    const prototype = Object.assign({}, this.__propertiesObject__, getWDIOPrototype('element'), { scope: 'element' })
+    const prototype = merge({}, this.__propertiesObject__, getWDIOPrototype('element'), { scope: 'element' })
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/browser/$.js]: same

File: packages/webdriverio/src/commands/element/$$.js [hunk 3]
```diff
@@ -39,14 +39,16 @@
...
-import { findElements, getPrototype as getWDIOPrototype, getElementFromResponse } from '../../utils'
+import {findElements, getBrowserObject, getPrototype as getWDIOPrototype, getElementFromResponse} from '../../utils'
 import { elementErrorHandler } from '../../middlewares'
 import { ELEMENT_KEY } from '../../constants'
 
 export default async function $$ (selector) {
     const res = await findElements.call(this, selector)
-    const prototype = Object.assign({}, this.__propertiesObject__, getWDIOPrototype('element'), { scope: 'element' })
+    const browser = getBrowserObject(this)
+    const prototype = merge({}, browser.__propertiesObject__, getWDIOPrototype('element'), { scope: 'element' })
```
Comment (christian-bromann) [inline packages/webdriverio/src/commands/element/$$.js]: same

File: packages/webdriverio/tests/addCommand.test.js [hunk 4]
```diff
@@ -47,6 +47,29 @@ describe('addCommand', () => {
...
+
+            browser.addCommand('myCustomElementCommand', async function () {
+                return this.execute(function () {return 1})
+            }, true)
+
+            const elem = await browser.$('#foo')
+
+            expect(await elem.myCustomElementCommand()).toBe(1)
+        })
+
```
Comment (christian-bromann) [inline packages/webdriverio/tests/addCommand.test.js]: These tests needs to move into the webdriver package, because you are testing the monad.

Comment (christian-bromann) [inline packages/webdriverio/tests/addCommand.test.js]: This should also fix the test coverage

File: packages/webdriverio/src/commands/browser/$$.js (hunk 1 above)
Comment (abjerstedt) [reply packages/webdriverio/src/commands/browser/$$.js]: _merge() still requires the result object to be the initial param.

File: packages/webdriverio/tests/addCommand.test.js (hunk 4 above)
Comment (abjerstedt) [reply packages/webdriverio/tests/addCommand.test.js]: Done - I didn't move, it... i wrote a separate test for monad.

Comment (abjerstedt) [reply packages/webdriverio/tests/addCommand.test.js]: It did!

**Comment by mgrybyk** [comment]
So now instead of $("body").addCommand(cmd, func) I should do browser.addCommand(cmd, func, true) ?

**Comment by abjerstedt** [comment]
@mgrybyk yep!

**Comment by mgrybyk** [comment]
Got it, I have to update typings then...

**Comment by abjerstedt** [comment]
@mgrybyk to elaborate, the previous method still works. In fact, i don't think we want people doing $('body').addCommand(name, func, true)... just:
browser.addCommand(name, func, true) or $('body').addCommand(name, func)

**Pull request merged**